doc/developer/design/20240411_graceful_reconfiguration_of_managed_clusters_v1.md

```diff
@@ -0,0 +1,364 @@
+off of the `sequence_staged` flow. In this scenario, we will still create new
+replicas in the foreground, but the wait/checks will be performed in a separate
+message handle by emitting a `ClusterReconfigureCheckReady` message.
+
+The state for the reconfiguration will largely be stored in a `PendingClusterReconfiguration`
+struct.
+
+
+```rust
+struct PendingClusterReconfiguration {
+ ctx: ExecuteContext,
+ // needed to determine timeouts
+ start_time: Instant,
+ // the plan will hold the cluster_id, config, mechanism, etc...
+ plan: AlterClusterPlan,
+ // idempotency key will need to be part of the  message and will need to be
+ // validated to ensure we're looking at the correct plan, in case we
+ // miss a cancel and a new entry was put in pending_cluster_reconfiguration.
+ idempotency_key: Uuid,
+}
+```
+
+In order to cancel we'll keep a map, `pending_cluster_reconfigurations`,  of
+session `ClusterIds` to `PendingClusterReconfigurations`.
+
+*Currently due to limitations in sources and sinks, we will need to maintain
+the disruptive alter mechanism. Once multiple replicas are allowed for sources
+and sinks we can remove the disruptive mechanism and default the wait to `WAIT=For 0 seconds`.
+
+
+#### Handling PendingClusterReconfigure Messages
+
+The responsibilities of this operation will be:
+1. checking for cancellations
+2. checking for the `WAIT` condition
+3. finalizaing the reconfiguration
+4. returning to the user
+
+__Checking the `WAIT` Condition__
+The message handler will perform all checks on every run. If the wait condition
+has not yet been met, we will reschedule the message to be resent. For each
+reschedule we'll spawn a task that waits some duration and sends a
+`ClusterReconfigureCheckReady` message with the connection_id.
+For `WAIT FOR` we can wait for the exact duration that would cause the next check to succeed.
+For `WAIT UNTIL CAUGHT UP` we can just wait for three seconds.
+
+__Finalize the Reconfiguration__
+Finalizing the reconfiguration will occur once the `wait` condition has been
+met. During this phase, the following steps will occur.
+- Drop any non pending replicas that do not have a billed_as value.
+- Finalize the pending replicas:
+  - Rename to remove the `-pending` suffix.
+  - Alter config to set `pending` to true.
+- Update the cluster catalog to the new desired config.
+
+__Handling Cancelation__
+When a cancel event is read we'll check to see if the active connection has a value
+in a `pending_cluster_reconfiguration` BTreeMap, and if so remove it and clean up leftover state.
+
+
+__v2 coniderations__
+Reverting the reconfiguration
+On failure or timeout of the reconfiguration, this task will revert to the state
+prior to the `ALTER CLUSTER` DDL. It will remove reconfigure replicas, and ensure the
+cluster config matches the original version.
+
+
+#### Recover from Environmentd Crash
+Initially, for v1, we will not recover. We will have to remove all `pending`
+replicas.
```

Comment:
```suggestion
#### Recover after environmentd restart
For v1, environmentd restarts will remove `pending` replicas from the Catalog during Catalog initialization.
This will cause the compute controller to remove any pending physical replicas during its initialization (this is already an existing feature).
```